### ISSUE TITLE
NC 20.0.2 Legacy encryption - mitigation and documentation thereof

### ISSUE THREAD

**Issue by Kuphi** [nextcloud/server#24240]
With Nextcloud 20, the term "Legacy encryption" was introduced. The documentation does say little about what legacy encryption is: https://docs.nextcloud.com/server/20/admin_manual/configuration_files/encryption_migration.html
I found another document, which hives hints about what could be meant: https://eprint.iacr.org/2020/1439.pdf

I run the command
`occ encryption:scan:legacy-format`
with output similar to this:
`sudo -u www-data php /var/www/nextcloud/occ encryption:scan:legacy-format
Scanning all files for legacy encryption
Scanning all files for username
/username/files_trashbin/files/file1.odt.d1592989576 does not have a proper header
/username/files_trashbin/versions/folder1/file2.xls.v1585723230 does not have a proper header
/username/files_versions/folder2/file3.xls.v1189677150 does not have a proper header`

Now I got some files with inproper headers. Does my nextcloud still use legacy encryption for all files or does it use a more secure encryption? What are the differences?

When running `occ encryption` I also encountered a new command:
`  Command "encryption" is not defined.

  Did you mean one of these?
      encryption:change-key-storage-root
      encryption:decrypt-all
      encryption:disable
      encryption:disable-master-key
      encryption:enable
      encryption:enable-master-key
      encryption:encrypt-all
      encryption:list-modules
      **encryption:migrate-key-storage-format**
      encryption:recover-user
      encryption:scan:legacy-format
      encryption:set-default-module
      encryption:show-key-storage-root
      encryption:status`

There is no documentation about **encryption:migrate-key-storage-format** but an excerpt from the file /nextcloud/core/Command/Encryption/MigrateKeyStorage.php - https://fossies.org/linux/nextcloud/core/Command/Encryption/MigrateKeyStorage.php
`75 ->setName('encryption:migrate-key-storage-format')
76 ->setDescription('Migrate the format of the keystorage to a newer format')`

Am I supposed to use this command? I have key-type: user keys ( https://docs.nextcloud.com/server/20/admin_manual/configuration_files/encryption_details.html#key-type-user-key )
Will I still be able to use user keys after running this command, and is it save to run it?

**Comment by Kuphi** [comment]
For the documentation record:
The command encryption:migrate-key-storage-format can not be run successfully, when your nextcloud is in maintenance mode. 
`
user@server:/home/user# sudo -u www-data php /var/www/nextcloud/occ encryption:migrate-key-storage-format
Nextcloud is in maintenance mode - no apps have been loaded
Updating key storage format
Start to update the keys:
    4 [----->----------------------]
In Manager.php line 187:
  Module with ID: OC_DEFAULT_MODULE does not exist.
encryption:migrate-key-storage-format
`
You have to turn off maintenance mode first:
`sudo -u www-data php /var/www/nextcloud/occ maintenance:mode --off`
then it runs smoothly:

`user@server:/home/user# sudo -u www-data php /var/www/nextcloud/occ encryption:list-modules`
`OC_DEFAULT_MODULE: Default encryption module [default*]`

`user@server:/home/user# sudo -u www-data php /var/www/nextcloud/occ encryption:migrate-key-storage-format`
`Updating key storage format`
`Start to update the keys:`
` 20 [============================]Key storage format successfully updated`

After the migration, **occ encryption:scan:legacy-format** does still output the same files as "does not have a proper header". Am I supposed to delete those files? Can I simply **rm** them and then do occ files:scan? (All my files with inproper header are either in the trashbin or in the folder files_versions.

**Comment by mark-orion** [comment]
Please someone explain what these "...does not have a proper header" error messages mean and how they can be fixed.

**Comment by leblitzdick** [comment]
> Please someone explain what these "...does not have a proper header" error messages mean and how they can be fixed.

I also have the same question, can i set "encryption.legacy_format_support" to false without any problems?

**Comment by BenedictBauerHHN** [comment]
I have the same output on occ encryption:scan:legacy-format of course to other files then the OP. Could it lead to problems if I set encryption.legacy_format_support to false?

**Comment by developerrespig** [comment]
I would also be interested in some more information about that issue from an official side. Besides of that I just had a look into the legacy format scanner where this message is originating from (you can find the source code of that checker here: https://github.com/nextcloud/server/blob/1448b7c923d079c9616f87df3ffa52a4656ac6cc/apps/encryption/lib/Command/ScanLegacyFormat.php (I guess)).

With my limited php skills I guess it looks like these messages mean, that the check failed (because the private method returns false and the expected "All scanned files are properly encrypted. You can disable the legacy compatibility mode." doesn't get rendered). Question for me would be now: What to do about them? As far as I have seen them on my end they are only related to file versions or trash bins. But for now I just leave that configurational entry in the config.php and I guess you should do the same until some says something different...

Hope to have something official here soon *fingers crossed*

**Comment by Kuphi** [comment]
**Can we get a comment regarding this issue?**

After upgrading to NC 20.0.3, it does anew complain about old server side encryption format. I thought I already resolved this?

"Das alte serverseitige Verschlüsselungsformat ist aktiviert. Wir empfehlen, es zu deaktivieren. Für weitere Einzelheiten sehe bitte in die Dokumentation."

**Comment by eckeSolutions** [comment]
In many instances the problem are old versions of files that are a few months old or even years.
Cleaning those versions with "occ versions:cleanup" removed all does “not have a proper header” errors for me. 
Proceed with care!!! Might delete necessary file versions!!!
running "occ encryption:scan:legacy-format" after that tells me it is save do set encryption.legacy_format_support to false

This is not a general solution, but should work for some people with this problem (if “/versions/” is included in the file urls).
A solution to set the proper header would be nice though.

**Comment by Kuphi** [comment]
> In many instances the problem are old versions of files that are a few months old or even years.
> Cleaning those versions with "occ versions:cleanup" removed all does “not have a proper header” errors for me.
> Proceed with care!!! Might delete necessary file versions!!!
> running "occ encryption:scan:legacy-format" after that tells me it is save do set encryption.legacy_format_support to false
> 
> This is not a general solution, but should work for some people with this problem (if “/versions/” is included in the file urls).
> A solution to set the proper header would be nice though.

Proceed with care? "occ versions:cleanup" does delete ALL previous versions of all files. This is only a solution if you do not need any previous versions at all.

**Comment by eckeSolutions** [comment]
> Proceed with care? "occ versions:cleanup" does delete ALL previous versions of all files. This is only a solution if you do not need any previous versions at all.
Yes, that’s why I added 3 exclamation marks. In my instance it was no problem. And to be on the save side I have daily backups of everything.

**Comment by DPTJKKVH** [comment]
It is really sad that nobody of the team even acknowledges that `encryption:migrate-key-storage-format` is undocumented and elaborates in a few words what it does and if we should do it after disabling legacy encryption.

Also @developerrespig seems to be correct. I removed the bad header files manually, ran the check, got the "you can disbale legacy encryption" message and did just that.

I see neither issues nor changes so far which is good. But it drives me nuts that I don't see the point and don't know if we should migrate the storage format because it seems to be security related.

@schiessle @rullzer @nickvergessen can you please explain if we should run `encryption:migrate-key-storage-format` or not after disabling legacy encryption, even if you don't feel to explain it any further? Just a YES or NO is sufficient for the time being.

Many thanks in advance!

**Comment by thomwiggers** [comment]
I don't think ``encryption:migrate-key-storage-format`` solves this problem. There doesn't seem to be any migration option if I look in the source code of the Encryption module.

**Comment by DPTJKKVH** [comment]
Well I can report that disabling legacy encryption breaks my installation, even though the encryption:scan:legacy-format tells me I can disable it.

To be more specific: Encrypted files cannot be downloaded anymore via web GUI and sharing files (e.g. publicly) doesn't work as well.

The GUI tells me that my private key is missing and I should log out and back in. But that doesn't fix it.

Then it tells me to set a new password. When changing paswords it tells me again that my private key is missing and I should log out and back in. Doing that again doesn't fix it.

So neither logging in nor changing passwords works.

Fortunately just enabling legacy encryption fixes the issue and I can access all files again. However this foreshadows doom for when legacy encryption won't be supported anymore. 

Either the scanning tool is broken or there is an important step missing after or before disabling legacy encryption.

I find it rather odd that simply disabling legacy encryption in the config is supposed to change encryption schemes. Also the "has been migrated" boolean in config.php is suggesting there might be more to it.

Oh btw: The encryption:migrate-key-storage-format does not run through and produces and error about an expexted boolean value for _decrypt_.

**Comment by thomwiggers** [comment]
It's not unlike that the files probably need to be migrated from the client side if you're using user key encryption. I only have a single user, and managed to migrate by `touch`-ing the few affected files. This triggered a re-upload in the client. This does however still leave old versions in `files_versions` that now show up in the scan tool.

> Either the scanning tool is broken or there is an important step missing after or before disabling legacy encryption.
I don't think the tool is broken, there simply isn't any migration path yet.

**Comment by DPTJKKVH** [comment]
> I don't think the tool is broken, there simply isn't any migration path yet.

That might be and it would be nice to have some official input on that.

The main issue I have at this point is that there is a scary orange warning in the admin panel that highly recommends admins to disable legacy encryption asap while at the same time it seems like there isn't a proper way to do that without breaking things.

I know everyone is de facto on leave for Christmas at this point and probably only paying customers get emergency support but one simple "Yea it's broken at the moment. Don't worry, we will fix it in 2021." would be nice or a little hotfix that simply removes the warning so nobody else falls victim to that would go a long way.

I know Nexctloud owes us nothing when we don't pay for support but it would be nice to not let people run into their doom right before Christmas.

**Comment by RedKage** [comment]
Hmmm
I have been using nextcloud for many years, migrated from owncloud 8 or something. So I must have started with nextcloud 10 or 11.
I have been using the default encryption module since then, and done web upgrades.

I have seem some posts somewhere about `encryption.legacy_format_support`, and I set it to `false`
Recently then I upgraded to 20.0.4.
Then emptied the trashbin.

Then I ran `encryption:scan:legacy-format` which took a while, and it told me a had a few files. they were all versions files.
I cleaned them up using `versions:cleanup`

Finally I have ran `encryption:migrate-key-storage-format` on version 20.0.4. 
Maintenance mode was activated while doing so
Nothing much was showing on stdout, and after about 1 hour it was done.
Everything is working properly, @DPTJKKVH I have had no weird things about changing password.

For me all is good, though I'm not sure anything was changed. I did huge backups before doing this, so in theory I could make some diffs to see if some stuff were actually changed... but, we are talking hundreds of gigs here.

**Comment by DPTJKKVH** [comment]
@RedKage I also run Nextcloud from the day there was a migration path from ownCloud.

What's interesting is that you ran key storage migration in maintenance mode. @Kuphi said it only worked when he disabled maintenance mode.

Also just for more background info: I had some files with bad/missing headers but none of them were versions or in the trash. I just backed them up on my client, removed them and emptied trash. Then I ran the legacy check and it said all is fine.

**Comment by Amon-san** [comment]
i also had this issue. i "fixed" it by copying the respective files to another directory, deleting the original and then moving the copy back in place. note: using 'move file' in the first place did not yield the same result for me.

**Comment by Kuphi** [comment]
> i also had this issue. i "fixed" it by copying the respective files to another directory, deleting the original and then moving the copy back in place. note: using 'move file' in the first place did not yield the same result for me.

Yes. By copying the files, they will lose all previous file versions and will be newly encrypted.

**Comment by Alvarord** [comment]
Hello, 

I face a similar issue as @DPTJKKVH 

> Well I can report that disabling legacy encryption breaks my installation, even though the encryption:scan:legacy-format tells me I can disable it.

I have upgraded from 19.0.7.1 to 20.0.5 yesterday

I cleaned the result from:

`
occ encryption:scan:legacy-format
`
When I deactivate the  _'encryption.legacy_format_support' => false_ then I can't access files and I see the #8546 Invalid private key. I tried [this](https://github.com/nextcloud/server/issues/8546#issuecomment-514139714) without success.

Then I change the value of  _'encryption.legacy_format_support' => true_  and I can access back to the data but the message *Invalid private key for encryption app. Please update your private key password in your personal settings to recover access to your encrypted files.* keeps appearing.

Not sure how can I help to track it.

I migrated from owncloud to nextcloud long ago (not sure in which version).

In the meantime I will do a full backup of my files (now that I can access them) just in case I break anything.

Any ideas about how to proceed?

More information:
```

root@machine:/var/www/html# sudo -u www-data php occ   encryption:scan:legacy-format
Scanning all files for legacy encryption
Scanning all files for user1
......
All scanned files are properly encrypted. You can disable the legacy compatibility mode.

root@machine:/var/www/html# sudo -u www-data php occ  encryption:status
  - enabled: true
  - defaultModule: OC_DEFAULT_MODULE

root@machine:/var/www/html# sudo -u www-data php occ  config:system:get version
20.0.5.2

```

**Comment by DPTJKKVH** [comment]
@Alvarord I ran encrypt-all multiple times and it doesn't fix anything. The issue as we experience it, stays the same.

I can only guess that people like us, who migrated from owncloud to nextcloud years ago, need some command to migrate everything encrypted in legacy mode to the new encryption, possibly including the new key storage format?

But alas two months go by and the developers have not even acknowledged that people have problems.

I fear for the worst when we have to update to v21 and our encryption breaks for good, forcing us to deploy from scratch, setup dozens of users and reupload gigabytes or terabytes worth of data.

**Comment by micah** [comment]
On 2021-01-19 06:51:47, DPTJKKVH wrote:
> I can only guess that people like us, who migrated from owncloud to
> nextcloud years ago, need some command to migrate everything encrypted
> in legacy mode to the new encryption, possibly including the new key
> storage format?

I did not upgrade from owncloud, and I had this problem.

**Comment by Alvarord** [comment]
Hello @DPTJKKVH , 

Is strange maybe they consider it should be solved with the encrypt-all, or they havent been able to reproduce it.

I saw in the bash history this morning that I executed the command yesterday but it was quite fast and I though it was because it was ciphering just a few old files. Today after doing the backup (luckily for me is only 100GB) I re-executed it and I noticed 2 things 

First it finish always with an error:
```

In Storage.php line 287:

  Could not decrypt key


In Crypto.php line 122:

  Authenticated ciphertext could not be decoded.
```
@DPTJKKVH do you have the same issue?


And every time I execute it it cipher just a few files less than 1000 files But at every try it is trying different groups of files (normally shared ones).

I will try to gather which file/s produce the issue.

Regards.

**Comment by DPTJKKVH** [comment]
@Alvarord you can't encrypt your files with encrypt-all because when you disable legacy encryption the server, for some reason, can't decrypt your private key. Try the same with legacy encryption enabled (just switch the "false" flag back to "true" in the config.php) and it will work.

Disabling legacy encryption does seemingly require the private keys to be stored in a different format. However `encryption:migrate-key-storage-format` interupts with an error. Maybe try this in both legacy and non legacy mode and tell us what your output is.

I think you and me have the same problem.


@micah do you HAVE the problem or HAD the the problem and fixed it?

**Comment by micah** [comment]
On 2021-01-19 07:27:40, DPTJKKVH wrote:

> @micah do you HAVE the problem or HAD the the problem and fixed it?

well, honestly I dont *know* if I have the problem still. I saw in my
administration panel that this was deprecated and I needed to run the
encryption:scan:legacy-format command. It produced a number of weird
errors and seemed like it failed. I tried moving those files out of the
way, like people here said, and re-ran it.

As far as I know, this just 'scanned' for files, and didn't do anything?
Will I need to do something more?

If I run this command and I don't get an error, do I no longer have the
problem?

When I run it now, it does not fail, and it says:

All scanned files are properly encrypted. You can disable the legacy compatibility mode.

How do I do that?

**Comment by Alvarord** [comment]
Hello @DPTJKKVH 

> @Alvarord you can't encrypt your files with encrypt-all because when you disable legacy encryption the server, for some reason, can't decrypt your private key. Try the same with legacy encryption enabled (just switch the "false" flag back to "true" in the config.php) and it will work.

same issue enabling or disabling it. A few files are encrypted and it finish with same error.

Trying right now the second part with legacy active (I think this took yesterday a while). I will try with both values and update you later.

Thanks for the ideas to try.

**Comment by DPTJKKVH** [comment]
@micah when it runs through without an error you are (in theory) ready to disable legacy encryption as described here:

https://docs.nextcloud.com/server/20/admin_manual/configuration_files/encryption_migration.html

Try that and then log into your normal user account on your nextcloud server and see if it tells you that your private key couldn't be decrypted. If that happens, you have the same issue as we have.

If you don't get any message, try downloading and opening your files via the web gui. If it works: Congratulations, you were lucky. If not, you suffer the same issue as we do.

**Comment by DPTJKKVH** [comment]
Please report your results when you are done. The more test cases we have, the better. @micah

**Comment by micah** [comment]
On 2021-01-19 08:09:05, DPTJKKVH wrote:
> @micah when it runs through without an error you are (in theory) ready to disable legacy encryption as described here:
>
> https://docs.nextcloud.com/server/20/admin_manual/configuration_files/encryption_migration.html
>
> Try that and then log into your normal user account on your nextcloud
> server and see if it tells you that your private key couldn't be
> decrypted. If that happens, you have the same issue as we have.

I set `encryption.legacy_format_support` to be `false` and then I logged
into nextcloud and it doesn't say anything...

I noticed there is also a setting `encryption.key_storage_migrated`,
this is currently set to `false`. Do I need to do something with this
too?

> If you don't get any message, try downloading and opening your files
> via the web gui. If it works: Congratulations, you were lucky. If not,
> you suffer the same issue as we do.

I am able to view the files, download and open them.

So I guess I'm lucky!

micah

**Comment by DPTJKKVH** [comment]
You are lucky then and I am more and more convinced that the errors @Alvarord and I have are due to us both truly coming from a full legacy encryption while most people already used the new encryption but with legacy support enabled.

**Comment by yahesh** [comment]
As the one who found that vulnerabilities that require those migration steps:

Having problems with the disabled legacy encryption generally means that you probably have files that encrypted but not integrity-protected. If your attack scenarios do not contain the modification of files then you don't have to fix this right away, but you should tackle this at some later point in time. For now you can set `encryption.legacy_format_support` to `false` and everything should work as before. Unless you know what you're doing, it's a **bad idea** to re-run the `encrypt-all` command. This will more likely break things instead of fixing them.

As for `encryption.key_storage_migrated`, there is another vulnerability that also through which the key material can be replaced. Do not just set this value, but instead use the `./occ encryption:migrate-key-storage-format` command.

**Comment by DPTJKKVH** [comment]
Hello @yahesh ! Really happy to have you step forward and give some context.

As described by @Alvarord and me, we did run `occ encryption:scan:legacy-format` and got errors about missing headers. Those were fixed by removing the files in question via normal means like moving them out of the sync folder and back into it. After that the command ran without issues and reported that we are ready to disable legacy encryption.

However when doing that in the config.php all files become unavailable through the web gui and a notification tells us that our private key could not be decrypted. There are two variants of this error notice: One tells you to log out and back in, which does not help. The other one tells you to change your password, which does not help either.

If I understand correctly the `encryption:migrate-key-storage-format` command is addressing a different issue and disabling legacy mode should work without it.

Still if I run this command either in maintenance mode or in production mode the command fails with a stout error. I hope I have it saved somewhere (will post it later) so I don't have to recreate it again and restore but either way this fails as well.

Are you sure that simply disabling legacy mode in the config.php magically changes the encryption of all files or do we rather have to re-encrypt them or do I miss something here?

Many thanks!

**Comment by Alvarord** [comment]
Hello, 
second part:

> Disabling legacy encryption does seemingly require the private keys to be stored in a different format. However `encryption:migrate-key-storage-format` interupts with an error. Maybe try this in both legacy and non legacy mode and tell us what your output is.

I didn't have any issue with the _migrate-key-storage-format_

`root@machine:/var/www/html/# sudo -u www-data php occ encryption:migrate-key-storage-format                                                                                                                                           
Updating key storage format
Start to update the keys:
    4 [============================]Key storage format successfully updated
`
With and without legacy active it works and everytime I try to encrypt all after with or without the legacy it fails after a few hundred files.

Seems that after the new re-run of the encrypt-all today I could login and remove the message about the private key properly (WEIRD maybe I didn't push the yes to accept the the first encryption?)

I was testing the download always with the same file and after the several tries with encrypt-all luckily I could download it without the legacy active and I can see in the traces the following:

`encrypt_all_at_13_31: encrypt files for user user1 (1 of 4): /user1/files/path/filetested.pdf
`
Seems the encrypt all make that file work now.

So seems my issue at this time is *JUST* in the encrypt-all that stops. I will try to make a script to make it run all night to check how many files it encrypt. As @yahesh points maybe is not a good idea (and I have no idea about what Im doing) but I think I have found something and I have backups so maybe it is possible to improve the error on the encrypt-all,  when the key fails there is no track about "what" fails, neither it checks the integrity of the information so maybe when you encrypt-all you should advice about integrity issues too. 

I have no idea about php but seems could be nice to add more info on this line 

https://github.com/nextcloud/server/blob/3cb8d85cf475061bcfbe6d901a9b9a0c29258ab4/lib/private/Encryption/Keys/Storage.php#L287

To check why the key decryption fail and maybe even it could be nice to catch it.

@DPTJKKVH you have problem with all the files? The encrypt all is succesful? Can you retry it?

**Comment by yahesh** [comment]
@DPTJKKVH As far as I remember there has been another issue in which people had the problem that the official command to scan for legacy encrypted files (`./occ encryption:scan:legacy-format`) would not find all occurences. I suggested to write a third-party script for that but have not yet come around to actually implementing it.

When the whole web UI breaks for you when disabling the legacy encryption this means that your private key is secured using the legacy encryption as well. Unfortunately, as far as I know, there is currently no built-in command to re-encrypt everything.

You understood correctly that the `./occ encryption:migrate-key-storage-format` command fixes a different vulnerability.

**Comment by DPTJKKVH** [comment]
@yahesh thank you so much for clarifying the part with the legacy encrypted private key. I am certain this must be my issue, probably in combination with not all legacy files being correctly detected.

@yahesh I pinged "the usual suspects" over a month ago but nobody seemed to be in inclined to even acknowledge the issue. If you have any way to get @schiessle @rullzer @nickvergessen or someone else's attention for this, it would be greatly appreciated. I do not use NC commercially but I run it for friends and family to keep them away from Dropbox, Google and so on (and I try to introduce it into the company I work for). It would be an atrocious time for everyone if I have to deploy a fresh Nextcloud instance, recreate all accounts, have everyone reupload their files (after having them back it up unencrypted locally) and everyone having to create dozens of new app tokens for their different devices.

I am thankful if you did build some unofficial tools to fix this but ultimately the team should create an official migration path, even if that path incorporates your code.

**Comment by yahesh** [comment]
@DPTJKKVH Do you know if you use user-based encryption keys? I would assume so because the master key encryption has been introduced much later. It might sufficient if all users changed their passwords. This would require Nextcloud to rewrite the private key files.

Concering the official command not finding all legacy encrypted file, my current guess is that those legacy encrypted private keys are the ones that aren't considered during the scan.

**Comment by DPTJKKVH** [comment]
@yahesh I do use user-based encryption and not by accident. I don't want to be able to decrypt user files even if I wanted to. :)

I already tested changing passwords both with my personal account, the admin users account (I have those seperated) and had another user change their password. In all instances the error that the private key could not be decrypted and the files being either unavailable or downloaded in encrypted state remained.

**Comment by DPTJKKVH** [comment]
Or should everyone change their password before disabling legacy mode and then I go through the motions of scanning etc.?

**Comment by yahesh** [comment]
@DPTJKKVH My assumption is that only the private key of the logged-in user will be decrypted. But those logic paths are outside of what I have poked around yet.

**Comment by DPTJKKVH** [comment]
@yahesh but I did understand you correctly: If changing passwords helps it must be done before disabling legacy encryption by every user, right? Because this is something I would love to try but I want to make sure I did not misunderstand you. ;)

Of course you assume no responsibility either way.

**Comment by yahesh** [comment]
@DPTJKKVH: Yes, each user would have to change their password once before disabling the legacy encryption. As the private keys are protected with the user passwords, they have to be rewritten when the password is changed. Unfortunately, I think this might not be enough, though, as the pubShare private key cannot be rewritten that easily.

**Comment by DPTJKKVH** [comment]
@yahesh is the pubShare key only for public shares (as the name suggests) or also used for shares between users of the same server?

**Comment by yahesh** [comment]
@DPTJKKVH They are only used for public shares. Sharing between users is done using the receiving user's public key.

**Comment by DPTJKKVH** [comment]
@yahesh many thanks for taking the time to explain and answer my questions. I will tell everyone to prepare for another try.

I will report back with results.

**Comment by Alvarord** [comment]
Hello, 

I want to point one important section in my previous comment:

> I was testing the download always with the same file and after the several tries with encrypt-all luckily I could download it **without** the legacy active and I can see in the traces the following:
> 
> encrypt_all_at_13_31: encrypt files for user user1 (1 of 4): /user1/files/path/filetested.pdf
> Seems the encrypt all make that file work now.

This means several things:
- The encryption:scan:legacy-format command is failing. Because it is not scanning all files properly or at least it is not detecting an issue in the file I mentioned.
- The encrypt-all command SHOULD be launched after the scan is successful and review integrity issues as mentioned in case #8546 

This is really **dangerous** because what I can see is that some files are properly accessible deactivating the legacy but some are not. This means that if the administrator trust the scan result it can **appear** that all files are correct without being true. Because maybe not all files had been encrypted properly.

As mentioned in #8546 my installation can be used at this time for experimentation if needed.

**Comment by DPTJKKVH** [comment]
To rule out any file modifications here are two Gist of both my ` MigrateKeyStorage.php` and `Crypto.php`.

https://gist.github.com/DPTJKKVH/88b203732550221740871a7eddb69e11

https://gist.github.com/DPTJKKVH/12504d857e215cf0beafc87100051946

**Comment by yahesh** [comment]
@DPTJKKVH I had a look at your stacktrace and I'd think that it shouldn't interrupt the execution of the key migration. When looking at [MigrateKeyStorage.php#L187](https://github.com/nextcloud/server/blob/master/core/Command/Encryption/MigrateKeyStorage.php#L187) one can see that the migration checks whether a key file has already migrated by trying to decrypt it. When the decryption succeeds, the migration process just skips the file. When the decryption fails, the corresponding Exception is cought and the migration process encrypts the key file.

**Comment by DPTJKKVH** [comment]
Hey @yahesh . Thanks for looking into it! 

So this is a whole new bug and I should probably open a new bug report, right?

**Comment by yahesh** [comment]
I'm not yet sure. Looking at the trace some more I stumbled over the `Argument 1 passed to OC\Security\Crypto::decrypt() must be of the type string, boolean given` message. It seems as if the call to `file_get_contents()` in [MigrateKeyStorage.php#L184](https://github.com/nextcloud/server/blob/master/core/Command/Encryption/MigrateKeyStorage.php#L184) fails and returns a boolean instead of the contents of the file. I assume this is not meant to happen.

**Comment by DPTJKKVH** [comment]
@yahesh I noticed that as well (see the first version of my post with the stacktrace) but edited it out because I didn't want to jump to conclusions.

Is there anything I can do to help debug this?

**Comment by yahesh** [comment]
@DPTJKKVH You could try to to `var_dump($path); var_dump($content);` after line 184 to find out which file path causes this problem and then see whether the corresponding file exists and is readable the system user that is used to migrate the keys.

**Comment by DPTJKKVH** [comment]
@yahesh Currently in the process of migrating the key storage format. Thanks to your suggestion I am able to see what files are causing issues but not why they cause issues. The stack trace for when an error occurs is identical to what we had before.

I'm currently running the migration command, wait for an error, let the user move out the files causing trouble, empty the trash bin and then re-run the migration command. I will repeat this process until it succeeds. Then I will report back. :)

**Comment by DPTJKKVH** [comment]
@yahesh Many thanks for all the support you are providing! I have some good news and some bad news:

Good: 

- After many iterations of the above described procedure the command ran through successfully. The `encryption.key_storage_migrated` entry in my config.php was automatically removed.
- I re-ran the `encryption:scan:legacy-format` command successfully and removed the corresponding entry in my config.php as well, effectively disabling legacy encryption. Then I restarted the apache2 process.
- None of the users gets an error message about their private key when they log in. So the migration seems to have indeed succeeded.
- Most of the files work properly. So they can be downloaded and opened fine.

Bad:

- Some files are broken. I was able to identify two kinds of errors:
- _Missing signature_ -> Results in an "File not available" error.
- _Encryption not ready: multikeydecrypt with share key failed:error:REDACTED_POSSIBLE_KEY_ID:rsa routines:RSA_padding_check_PKCS1_type_2:pkcs decoding error_ -> Results in an "Internal Server Error".

None of the affected files are shared, so the second error is especially confusing.

I log Nextcloud errors to syslog so here are the corresponding error messages from the syslog as gists:

[Signature Error](https://gist.github.com/DPTJKKVH/36f670ba610d03e1cc943cdf51f63b18)

[Multi Key Decrypt Error](https://gist.github.com/DPTJKKVH/c8357dcd1ed8e2b5fa0f90fc2755a91b)


One additional question: As you can see I had to redact a `private key` that was written to syslog when the second error occurred. That sounds rather dangerous to me. You don't have to explain it to me but I wanted to notify you of this because you are able to determine if that is a security issue which should be reported.

**Comment by DPTJKKVH** [comment]
@yahesh I figured that you won't be able to help me with the remaining issues (signature and multikeydecrypt) and with some testing I noticed that I can simply remove the "broken" files and reupload them to fix my remaining issues.

However, please do look at the logs about the [Multi Key Decrypt Error](https://gist.github.com/DPTJKKVH/c8357dcd1ed8e2b5fa0f90fc2755a91b) since I am certain that private keys should never be written to a log file. I trust you can determine if this is something that should be reported as a security vulnerability.

**Comment by yahesh** [comment]
@DPTJKKVH I had a look at the Gist and the place where the private key is written to the logfile seems to be in [Crypt.php#L683](https://github.com/nextcloud/server/blob/fda6ffc866cf8c5d3579fe95d1731ab747894002/apps/encryption/lib/Crypto/Crypt.php#L683). The error logging seems to be so generic that it also outputs arguments while, in this case, it leads to a security vulnerability as the plaintext private key used to the decrypt the files is printed to logfile as well. IMHO this should definitely be reported to the maintainers as a security vulnerability.

There are more occurences where a thrown exception might lead to leaked secret information, e.g. in [Crypt.php#L505](https://github.com/nextcloud/server/blob/fda6ffc866cf8c5d3579fe95d1731ab747894002/apps/encryption/lib/Crypto/Crypt.php#L505), [Crypt.php#L617](https://github.com/nextcloud/server/blob/fda6ffc866cf8c5d3579fe95d1731ab747894002/apps/encryption/lib/Crypto/Crypt.php#L617) and [Crypt.php#L677](https://github.com/nextcloud/server/blob/fda6ffc866cf8c5d3579fe95d1731ab747894002/apps/encryption/lib/Crypto/Crypt.php#L677).

**Comment by DPTJKKVH** [comment]
@yahesh I feel really bad to ask this from you after you helped me so much already but would you like to report this security issue to Nextcloud? I don't know how I would go about to properly report such an issue other than pointing to our conversation here. You on the other hand have already identified multiple additional instances where such an issue could occur as well.

The fact that private keys can be exposed by such trivial means is extremely scary. I guess this is an inherent problem when encryption happens server side instead of client side? However the least that should be done is limiting default error logging in a way that confidential information is not exposed.

It's kind of ironic that the Signature error log does contain instances of `*** sensitive parameter replaced ***`. Those are not done by me manually but by Nextcloud automatically...

**Comment by Alvarord** [comment]
Hello, 

just to leave it written, the encrypt-all finish without error now that I tracked the exception with a var_dump(path) in front of:

https://github.com/nextcloud/server/blob/3cb8d85cf475061bcfbe6d901a9b9a0c29258ab4/lib/private/Encryption/Keys/Storage.php#L287)

> First it finish always with an error:
> 
> ```
> 
> In Storage.php line 287:
> 
>   Could not decrypt key
> 
> 
> In Crypto.php line 122:
> 
>   Authenticated ciphertext could not be decoded.
> ```

It happened with a shared folder with just more folder in it and no files. So seems nextcloud was expecting data inside. I have removed the folder and the encrypt-all finish properly without exception **BUT** (and it is a big one) it still keeps encrypting not all files and each time I execute it is choosing a different subset of files.

I will try to check how the subset is chosen, at this time i'm trying to understand this part:

https://github.com/nextcloud/server/blob/3cb8d85cf475061bcfbe6d901a9b9a0c29258ab4/apps/encryption/lib/Crypto/EncryptAll.php#L264

There is where the loop cipher all files happens but I don't understand why it finish early in my case, I can see the $directories has a lot of objects with a var_dump but there is not the same amount of messages with the encrypt files and there is no log with files already encrypted neither.

Im not sure how to track this any further as from my point of view the code is totally correct but is strange that the loop for cphering the user finish early but I can see content in the directories and even I see a lot of files added so the getDirecrotyContent seems to be working properly too. It doesnt make any sense.

I will keep working on it.

If anybody has any idea please advice me.

Regards.

**Comment by Alvarord** [comment]
Hello, 

the issue with my file is exactly this one reported by @DPTJKKVH 
 
> [Signature Error](https://gist.github.com/DPTJKKVH/36f670ba610d03e1cc943cdf51f63b18)

So that seems definitively related to the issue #8546 with the 00sig00

**Comment by Cardes** [comment]
Hello,

first: thanks for all your helpfull comments here, those information helped me mitigate a lot of errors while migrating to the new encryption format. Sadly some obstacles still remain:

Maybe this is related and somewhat important if you are using the collabora online app (or try to use):
The reason I'm writing this here is because it happens on files created via web interface **after** switching to the new encryption.

I get "Legacy cipher is no longer supported!" while using the collabora app for newly created files where everything else is working fine. 
Maybe its just not supported but it seems to mee the app is still somehow creating files with the old encryption format while everything else is using the new format. 
For reference: 
I migrated all users with password change in advance (coming from age old owncloud, migrated years ago to nextcloud), cleaned up the trash and versions files.
I got no errors using scan:legacy-format & doing key-migration
Users see no error message in the gui about invalid private key and can access their files just fine, create new ones and so on.
I can reliably reproduce the behaviour with the collabora app which is just working fine on unencrypted, external storage shares using smb.

**first variant with empty file**
1. create an empty odt file in the users home folder or subfolder via web/files app "+"
2. access with collabora is fine and the file can be saved
3. reopen the file, edit some text and try to save: error is logged: "Legacy cipher is no longer supported!" and "Private Key missing for user: please try to log-out and log-in again" also showing an error in collabora 

**second variant with template file**
1. create a odt file from template
2. open the file in collabora and try to safe it: same errors as above
3. 
The difference between both situations seems to be that an empty file can be successfully written once while an existing file cannot be edited. 

I realize I'm in a somewhat special situation where i use unencrypted, external storage via smb which just works fine and encrypted home folders for users where the error happens but I still whanted to create some awarness of this situation because I feel some people might too get stuck in a situation like me given that i could't find something in the documentation about the collabora app about encryption is unsupported.

**Comment by kaysond** [comment]
I'm curious if anyone has run into the problem where `occ encryption:scan:legacy-format` returns files/folders that no longer exist?

Then when I run `occ encryption:migrate-key-storage-format`, it crashes with the same message as in https://github.com/nextcloud/server/issues/24240#issuecomment-764703146 because the files don't exist.

It seems that `$this->rootView->getDirectoryContent($folder)` is traversing something that's not actually the directory structure on disk... Maybe there's some kind of structure I'm missing... I'm still working through the files to debug, but any tips would be appreciated!

**Comment by yahesh** [comment]
Might be that it takes the file list in the database as the basis for the file scan.

**Comment by kaysond** [comment]
@yahesh looks like you might be right. I turned off the db, and it errors out almost immediately, though its happening in `OC::init`, which may not necessarily be the code that traverses the directory. I'll open a separate issue for this

**Comment by eNTi** [comment]
I'm seeing the same issue. Getting lots of 

> does not have a proper header

messages for files. No idea what to do about it and how to proceed. Can I somehow resync the db so that it sees the correct files? And if so... what's next? Can I change the encryption for the files on the server for all users? So are they in effect getting decrypted and encrypted again? Is this dangerous?

Some guidance would be much appreciated!

**Comment by DPTJKKVH** [comment]
Hello everyone. Sad to see that the issue isn't even acknowledged by the devs yet. If it wasn't for Yahesh (🙏) I would have been still stuck as well.

I don't have a nice solution for you guys but I can reiterate how I "fixed" the issue on my end.

Basically I undid all changes I made in trying to debug the issue. So I had a "vanilla" legacy encryption setup. Then I told every user of my server to locally backup their files and the delete everything stored on Nextcoud, including wiping the versions and trash bin. Then I migrated the key storage format and disabled legacy encryption. After that everyone had to reupload all of their files.

It was a painful process that took severals days but after that everything was well and still is well for months. I strongly suggest to go this route since I honestly exhausted every other possibility described here and beyond and none worked.

Only public shares do not work anymore and I guess that is caused by the new key storage format. It's annoying but after the trouble everyone went through we all decided to rather deal with it than risk breaking anything.

All in all this was an atrocious experience and I'm glad that nobody threw the towel and decided to sell out to Google or DropBox.

Best of luck!

**Comment by jknockaert** [comment]
Using NC encryptions has been a pita for a long time now. The implementation has been a mess from the very beginning (even before NC forked from OC), and documentation has been largely absent or even misleading (remember #8311). I've been trying to keep it going, starting with reimplementing the encryption stream wrapper (the original one being terribly inefficient) and then later fixing errors here and there. Often new errors came with new versions of NC, resulting from largely undocumented commits. PRs to fix them sometimes take several months after submission before receiving a review. Enough's been enough, so today I decided to go without encryption and I switched it off. I will no longer be available to review or fix any issues with respect to encryption in NC.

**Comment by szaimen** [comment]
@jknockaert That's comprehensible but of course sad to hear!

> Enough's been enough, so today I decided to go without encryption and I switched it off

I wonder what you use instead? No encryption at all or e.g. FDE?

**Comment by jknockaert** [comment]
> I wonder what you use instead? No encryption at all or e.g. FDE?

I started using ZFS native encryption.

**Comment by J0WI** [comment]
Unfortunately, FS encryption can't replace user key encryption.

**Comment by DPTJKKVH** [comment]
Guys please. I do share your frustration 100% but this issue is for debugging and supporting people to get through the migration process and not for discussing alternatives to the standard encryption module.

This issue is very crowded and you ping everyone with your private conversation.

**Comment by szaimen** [comment]
<details>
<summary>OT @J0WI</summary>

> Unfortunately, FS encryption can't replace user key encryption.

I am not sure if it was actually ever meant to be really secure from the admins: 
Citing this doc https://github.com/nextcloud/desktop/blob/master/doc/architecture.rst#end-to-end-encryption:
> While data on the server can be encrypted, this is largely designed to protect it from malicious storage solutions or theft of the whole hardware. System administrators always have access to the data.

</details>

**Comment by eNTi** [comment]
Not to be ungrateful but @DPTJKKVH is right. Can you move this conversation somewhere else please?

**Comment by jknockaert** [comment]
The fun thing is, even after disabling encryption (`occ encryption:disable`) I still get the message:
> The old server-side-encryption format is enabled. We recommend disabling this. For more details see the documentation.

That's quite a bit off. And don't expect the documentation to help you any further; it rarely did when it comes to the encryption app.

**Comment by kaysond** [comment]
https://github.com/nextcloud/server/issues/24240#issuecomment-819264419
So the solution to this, in my case, was to run `occ files:scan --all` (requires maintenance mode being **off**). Then it ran with no issues!

**Comment by DPTJKKVH** [comment]
@kaysond I don't think it's related to the issue at hand. I suddenly have gotten such errors as well, after the server ran more or less okay for a time. I set up a second server, 100% fresh via the Snap repository, no backup restore or anything. I enabled the encryption module uploaded some files, and there were the errors already.

This is where I gave up. We are on SeaFile now.

But again: I don't think it's related to the migration since the errors happen on a fresh install as well, right after enabling the default encryption module.

**Comment by kaysond** [comment]
> 
> 
> @kaysond I don't think it's related to the issue at hand. I suddenly have gotten such errors as well, after the server ran more or less okay for a time. I set up a second server, 100% fresh via the Snap repository, no backup restore or anything. I enabled the encryption module uploaded some files, and there were the errors already.
> 
> This is where I gave up. We are on SeaFile now.
> 
> But again: I don't think it's related to the migration since the errors happen on a fresh install as well, right after enabling the default encryption module.

Thanks for the info. I wonder if it is just the upgrade to the new version that caused this.

**Comment by szaimen** [comment]
Maybe this comment shelds some lights on something here?
https://github.com/nextcloud/server/issues/27501#issue-920474958

**Comment by ajs124** [comment]
In my case, the files are just plain unencrypted files, in the filesystem. Nextcloud however seems to think that they are encrypted. I've already run `occ files:scan --all` and so on, but `encryption:scan:legacy-format` still finds the same files.

I fixed this by deleting the folders and files relating to those files from `files_encryption/keys`, dropping the entries referencing them from the database and setting the `encrypted` field to `0` in the database. Then I ran `files:scan --all` again and now it seems to work. Idk if that helps anyone here.

**Comment by PVince81** [comment]
for the case of unencrypted files wrongly detected as encrypted, please run: https://docs.nextcloud.com/server/latest/admin_manual/issues/general_troubleshooting.html#problems-when-downloading-or-decrypting-files

**Comment by Gajusbonus** [comment]
> for the case of unencrypted files wrongly detected as encrypted, please run: https://docs.nextcloud.com/server/latest/admin_manual/issues/general_troubleshooting.html#problems-when-downloading-or-decrypting-files

Returns:                                                              
  Command "encryption:fix-encrypted-version" is not defined.
occ config:system:get version
20.0.10.1
So....what version do you expect me to run, to be able to fix decrypting files?

**Comment by szaimen** [comment]
Hi, please update to at least 23.0.12 and report back if it fixes the issue. Thank you!

**Comment by RedKage** [comment]
Do you guys a favor and stop using encryption. This is broken since I have ever used nextcloud.
And started with version 9 or 10.
All these years I had 500 errors with bad encryption errors, had to re-up files to fix these issues.
For 4 years I coped with this, until version 21 or something.

Then decided to reinstall everything, re-upload everything without encryption and do my own encryption with gocrypt-fs. I'm never going back. This is a broken feature from the start. And also prevent many apps to work with it, like the music app which tries (tried? Maybe fixed) to save tags on encrypted files and corrupt them or many other apps not handling encryption.
Also now that I have removed encryption it's very easy for to backup and restore my data, as the data is unencrypted I can pull files directly from a backup. While still benefiting from fs encryption on top. 

Anyway my 2cents guys.

**Comment by Pazu** [comment]
Hear, hear. The worst part is there is no apparent will or perhaps even ability on the part of the developers to do anything about this problem.

**Comment by asheroto** [comment]
This issue is 3 years old. Even on a brand new nextcloud installation the issue exists. If it's broken, it should be disabled and not offered as a working product.

Going to try the solution @RedKage proposed.

**Comment by Flonky** [comment]
Hi, running `./occ encryption:fix-encrypted-version user` seems to have fixed this for me, although I also had to run `./occ files:scan --all` afterwards.

For reference, `occ config:system:get version` returns `28.0.1.1`.

**Comment by asheroto** [comment]
Indeed, it will appear to fix it at first. But I found several days later that encryption was still broken for new or changed files moving forward, especially larger than a few KB even though running that command fixes the existing files.

This of course may not be the case for everyone, but when using the previously mentioned encrypted method encryption with S3, I found this to be the case.

**Comment by asheroto** [comment]
> Even after key migration? I haven't noticed anything like that yet, although my instance doesn't have too much traffic either.

Yes! Even after that. I worked on this for days and made several posts. :-/ I ended up just switching to [SSE-C](https://github.com/nextcloud/server/issues/33283) with my S3 provider. That worked out fine, although it still rubs me the wrong way that my data is sent to the S3 provider without being encrypted first. I realize it is encrypted in transmit, but the S3 provider technically _receives_ the raw data, _then_ encrypts it at rest. 😊

**Comment by brotkastn** [comment]
Hello everyone.
Since there have been some time since the last comment here, i thought i would take my chance again at disabling the legacy encryption.

Previously, i did the `encryption:migrate-key-storage-format` . Today i have taken the following steps:

```
www-data@srv-nextcloud:~/nextcloud$ php occ status
  - installed: true
  - version: 29.0.4.1
  - versionstring: 29.0.4
  - edition: 
  - maintenance: false
  - needsDbUpgrade: false
  - productname: Nextcloud
  - extendedSupport: false
www-data@srv-nextcloud:~/nextcloud$ php occ files:scan --all
www-data@srv-nextcloud:~/nextcloud$ php occ encryption:scan:legacy-format | tee ../../nextcloud-datadir/nextcloud_encryption_legacy_2024-07-25
Scanning all files for legacy encryption
Scanning all files for brot
Scanning all files for otheruser1
Scanning all files for otheruser2
Scanning all files for otheruser3
Scanning all files for otheruser4
Scanning all files for otheruser5
All scanned files are properly encrypted. You can disable the legacy compatibility mode.
```

However, if i try to access a share my wife made with a collection of family photos, via the link she used to send to our extended family, it seems to miss the legacy encryption. I can view the previews of the photos, but nextcloud throws an error when i try to access stuff. Neither the "Download" button on the top of the page or the download of an individual image seems to work. The nextcloud log shows the following entries:

```
{"reqId":"MSajkOfIBJz2hC2BEbwH","level":3,"time":"2024-07-25T18:10:54+00:00","remoteAddr":"2001:1234::1234","user":"--","app":"no app in context","method":"GET","url":"/s/shareViaLink/download","mes
sage":"Exception thrown: OC\\ServerNotAvailableException","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:128.0) Gecko/20100101 Firefox/128.0","version":"29.0.4.1","exception":{"Exception":"OC\\ServerNotAvailab
leException","Message":"Legacy cipher is no longer supported!","Code":0,"Trace":[{"file":"/var/www/nextcloud/apps/encryption/lib/Crypto/Crypt.php","line":378,"function":"getLegacyCipher","class":"OCA\\Encryptio
n\\Crypto\\Crypt","type":"->"},{"file":"/var/www/nextcloud/apps/encryption/lib/KeyManager.php","line":403,"function":"decryptPrivateKey","class":"OCA\\Encryption\\Crypto\\Crypt","type":"->","args":["*** sensiti
ve parameters replaced ***"]},{"file":"/var/www/nextcloud/apps/encryption/lib/Crypto/Encryption.php","line":155,"function":"getFileKey","class":"OCA\\Encryption\\KeyManager","type":"->"},{"file":"/var/www/nextc
loud/lib/private/Files/Stream/Encryption.php","line":284,"function":"begin","class":"OCA\\Encryption\\Crypto\\Encryption","type":"->"},{"function":"stream_open","class":"OC\\Files\\Stream\\Encryption","type":"-
>"},{"file":"/var/www/nextcloud/lib/private/Files/Stream/Encryption.php","line":212,"function":"fopen"},{"file":"/var/www/nextcloud/lib/private/Files/Stream/Encryption.php","line":187,"function":"wrapSource","c
lass":"OC\\Files\\Stream\\Encryption","type":"::"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":489,"function":"wrap","class":"OC\\Files\\Stream\\Encryption","type":"::"}
,{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Wrapper.php","line":307,"function":"fopen","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/F
iles/View.php","line":1169,"function":"fopen","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/View.php","line":997,"function":"basicOperation","class":"
OC\\Files\\View","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Node/File.php","line":116,"function":"fopen","class":"OC\\Files\\View","type":"->"},{"file":"/var/www/nextcloud/lib/private/Streamer.p
hp","line":132,"function":"fopen","class":"OC\\Files\\Node\\File","type":"->"},{"file":"/var/www/nextcloud/lib/private/legacy/OC_Files.php","line":217,"function":"addDirRecursive","class":"OC\\Streamer","type":
"->"},{"file":"/var/www/nextcloud/apps/files_sharing/lib/Controller/ShareController.php","line":492,"function":"get","class":"OC_Files","type":"::"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dis
patcher.php","line":232,"function":"downloadShare","class":"OCA\\Files_Sharing\\Controller\\ShareController","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":138,"fu
nction":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/App.php","line":184,"function":"dispatch","class":"OC\\AppFramework\\Ht
tp\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/Route/Router.php","line":338,"function":"main","class":"OC\\AppFramework\\App","type":"::"},{"file":"/var/www/nextcloud/lib/base.php","line":
1050,"function":"match","class":"OC\\Route\\Router","type":"->"},{"file":"/var/www/nextcloud/index.php","line":49,"function":"handleRequest","class":"OC","type":"::"}],"File":"/var/www/nextcloud/apps/encryption
/lib/Crypto/Crypt.php","Line":301,"CustomMessage":"Exception thrown: OC\\ServerNotAvailableException"}}
{"reqId":"AjIzOqiDuIKw0n44A543","level":3,"time":"2024-07-25T18:11:11+00:00","remoteAddr":"2001:1234::1234","user":"--","app":"no app in context","method":"GET","url":"/s/shareViaLink/download?path=
&files=singleImage.jpg","message":"Exception thrown: OC\\ServerNotAvailableException","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:128.0) Gecko/20100101 Firefox/128.0","version":"29.0.4.1","excep
tion":{"Exception":"OC\\ServerNotAvailableException","Message":"Legacy cipher is no longer supported!","Code":0,"Trace":[{"file":"/var/www/nextcloud/apps/encryption/lib/Crypto/Crypt.php","line":378,"function":"
getLegacyCipher","class":"OCA\\Encryption\\Crypto\\Crypt","type":"->"},{"file":"/var/www/nextcloud/apps/encryption/lib/KeyManager.php","line":403,"function":"decryptPrivateKey","class":"OCA\\Encryption\\Crypto\
\Crypt","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"/var/www/nextcloud/apps/encryption/lib/Crypto/Encryption.php","line":466,"function":"getFileKey","class":"OCA\\Encryption\\KeyManag
er","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Storage/Wrapper/Encryption.php","line":354,"function":"isReadable","class":"OCA\\Encryption\\Crypto\\Encryption","type":"->"},{"file":"/var/www/nex
tcloud/lib/private/Files/Storage/Wrapper/Wrapper.php","line":182,"function":"isReadable","class":"OC\\Files\\Storage\\Wrapper\\Encryption","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/View.php","l
ine":1171,"function":"isReadable","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/View.php","line":491,"function":"basicOperation","class":"OC\\Files\\V
iew","type":"->"},{"file":"/var/www/nextcloud/lib/private/Files/Filesystem.php","line":530,"function":"isReadable","class":"OC\\Files\\View","type":"->"},{"file":"/var/www/nextcloud/lib/private/legacy/OC_Files.
php","line":344,"function":"isReadable","class":"OC\\Files\\Filesystem","type":"::"},{"file":"/var/www/nextcloud/lib/private/legacy/OC_Files.php","line":123,"function":"getSingleFile","class":"OC_Files","type":
"::"},{"file":"/var/www/nextcloud/apps/files_sharing/lib/Controller/ShareController.php","line":487,"function":"get","class":"OC_Files","type":"::"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dis
patcher.php","line":232,"function":"downloadShare","class":"OCA\\Files_Sharing\\Controller\\ShareController","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":138,"fu
nction":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/AppFramework/App.php","line":184,"function":"dispatch","class":"OC\\AppFramework\\Ht
tp\\Dispatcher","type":"->"},{"file":"/var/www/nextcloud/lib/private/Route/Router.php","line":338,"function":"main","class":"OC\\AppFramework\\App","type":"::"},{"file":"/var/www/nextcloud/lib/base.php","line":
1050,"function":"match","class":"OC\\Route\\Router","type":"->"},{"file":"/var/www/nextcloud/index.php","line":49,"function":"handleRequest","class":"OC","type":"::"}],"File":"/var/www/nextcloud/apps/encryption
/lib/Crypto/Crypt.php","Line":301,"CustomMessage":"Exception thrown: OC\\ServerNotAvailableException"}}
{"reqId":"AjIzOqiDuIKw0n44A543","level":3,"time":"2024-07-25T18:11:11+00:00","remoteAddr":"2001:1234::1234","user":"--","app":"PHP","method":"GET","url":"/s/shareViaLink/download?path=&files=singleImage.jpg","message":"Undefined variable $event at /var/www/nextcloud/lib/private/legacy/OC_Files.php#242","userAgent":"Mozilla/5.0 (X11; Linux x86_64; rv:128.0) Gecko/20100101 Firefox/128.0","versio
n":"29.0.4.1","data":{"app":"PHP"}}
```

I do wonder why that exact folder still relies on legacy encryption, as said pictures have been uploded around may - the "active" encryption-module was already set to the default one.

```
www-data@srv-nextcloud:~/nextcloud$ php occ encryption:show-key-storage-root
Current key storage root:  default storage location (data/)
www-data@srv-nextcloud:~/nextcloud$ php occ encryption:status
  - enabled: true
  - defaultModule: OC_DEFAULT_MODULE
```

I wonder what the way forward to disabling the legacy encryption module is. The  re-download and start with a fresh nextcloud approach is a bit problematic since we are talking about 1.2TB worth of files, family spread across a few thousand kilometers and a huge number of share links which would be lost:

```
MariaDB [nextcloud]> select count(*) from oc_share where token IS NOT NULL;
+----------+
| count(*) |
+----------+
|      391 |
+----------+
1 row in set (0.002 sec)
```

Thanks for reading. I am happy that the community proved so helpful for people trying to get their data back. If there is someone with an idea how to proceed in my case, i am happy to test even more 'creative' solutions to this problem and report back.

Also, thanks to the devs for the awesome suite of software that nextcloud is :)
